scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -384,26 +397,34 @@ public boolean advance() throws IOException {
         }
 
         int keyGroupCount = 0;
+
+        // Track the canonical # buckets of each source that the key is found in.
+        // If we find it in a source with a # buckets >= the parallelism of the job,
+        // we know that it doesn't need to be re-hashed as it's already in the right bucket.
+        boolean keyFoundInSourceWithMatchingNumBuckets = false;
```

Comment:
this is actually a generalization of the previous heuristic used to determine if a key needs re-hashing. Consider this example of a LHS with 1 bucket and a RHS with 2, where the parallelism is set to 2 (so number of buckets in output == 2).

```
Source A    | Source B               
Bucket 0    |  Bucket 0   |  Bucket 1
-------------------------------------
1           |  1          | 4
2           |  3
3                
4
```

Worker 1 is merging `SourceA-Bucket0` with `SourceB-Bucket0`. When we come across element `1`, we see it exists in Source B. Since source B has 2 buckets, which == the parallelism of the source, we know that if we performed a re-hashing of `1`, it would still be assigned to bucket 0 in a 2-bucket sink. On the other hand, when we get to element `2`, since it does not exist in Source B, we don't know where it would end up in a 2-bucket sink operation, so we must rehash.

Worker 2 is merging a copy of `SourceA-Bucket0` with `SourceB-Bucket1`. It would need to rehash 1, 2, and 3, but not 4. 

In datasets with fanout and a lot key overlap (like event data!) this should eliminate a lot of hashing operations.

@nevillelyh lmk if you can poke any holes in this idea, I tested it pretty thoroughly (including with partitioned sources where different partitions have different # buckets), but could be missing something

Reply:
Just thinking out loud: when `numBuckets` of a source >= parallelism, its pairs are read in a merge sort of 1 or more buckets, so any key in there are in the right place. We have to check keys from other sources (replicated to increase parallelism) whether they belong here. Did I get this right?

Reply:
Yep, exactly